python/GafferUI/ColorChooserPlugValueWidget.py

```diff
@@ -89,6 +126,34 @@ def __colorChanged( self, colorChooser, reason ) :
 				for plug in self.getPlugs() :
 					plug.setValue( self.__colorChooser.getColor() )
 
+	def __colorChooserOptionChanged( self, colorChooser, value, key ) :
+
+		if Gaffer.Metadata.value( "colorChooser:inlineOptions", "userDefault" ) is None :
+			sessionOptions = Gaffer.Metadata.value( "colorChooser:inlineOptions", "sessionDefault" )
+			if sessionOptions is None :
+				sessionOptions = IECore.CompoundData()
+				Gaffer.Metadata.registerValue( "colorChooser:inlineOptions", "sessionDefault", sessionOptions )
+
+			sessionOptions.update( { key: value } )
+
+		for p in self.getPlugs() :
+			plugOptions = Gaffer.Metadata.value( p, "colorChooser:inlineOptions" )
+			if plugOptions is None :
+				plugOptions = IECore.CompoundData()
+				Gaffer.Metadata.registerValue( p, "colorChooser:inlineOptions", plugOptions, persistent = False )
+
+			plugOptions.update( { key: value } )
+
+	def __colorChooserOptions( self ) :
+
+		v = sole( Gaffer.Metadata.value( p, "colorChooser:inlineOptions" ) for p in self.getPlugs() )
+		if v is None :
+			v  = Gaffer.Metadata.value( "colorChooser:inlineOptions", "userDefault" )
+			if v is None :
+				v = Gaffer.Metadata.value( "colorChooser:inlineOptions", "sessionDefault" ) or IECore.CompoundData()
```

Comment:
I'm not sure what this provides on top of the usual metadata inheritance rules either - is there something I'm missing?

Taking on board recent comments about startup files being a bit much for a non-technical standalone user, I wonder if we should add a "Save Defaults" menu item that creates `~/gaffer/startup/gui/colorChooser.py` and makes the `Color[34]fPlug` registrations based on the values from the current plug?

Reply:
My goal was to treat `userDefault` (I see your point on registering to `Color[3,4]fPlug` so I'll refer to that from now on) as the behavior the user wants regardless of what they change during a session. I figured registering metadata in a startup script is a very proactive choice and we should stick to that throughout a session, regardless of what a user changes during a session.

To accomplish that I needed the two separate metadata entries so there's one entry that gets updated as the user makes choices (unless there's a setting from a startup script) and (possibly) one that is kept static because it was set in a startup script.

But maybe that behavior is not the best, or too complicated? If we always register a choice to a `Color3fPlug`, users can set options they get at startup, and then if they change an option, that will become the new default for `ColorChooser` without a plug-specific registration.

Reply:
This was improved in the great Color Chooser metadata overhaul of 2024: https://github.com/GafferHQ/gaffer/pull/6003/commits/8bf9543bafcf26dd41e265f7f18376a2d8460a4a. Now we have an explicit method of saving defaults, including creating a startup script to load the settings.